astropy/utils/data.py

```diff
@@ -1375,15 +1714,103 @@ def get_cached_urls():
     -------
     cached_urls : list
         List of cached URLs.
-    """
-    # The code below is modified from astropy.utils.data.download_file()
-    try:
-        dldir, urlmapfn = _get_download_cache_locs()
-    except OSError as e:
-        msg = 'Remote data cache could not be accessed due to '
-        estr = '' if len(e.args) < 1 else (': ' + str(e))
-        warn(CacheMissingWarning(msg + e.__class__.__name__ + estr))
-        return False
 
-    with shelve.open(urlmapfn) as url2hash:
+    See Also
+    --------
+    cache_contents : obtain a dictionary listing everything in the cache
+    """
+    with _cache() as (dldir, url2hash):
         return list(url2hash.keys())
+
+
+def cache_contents():
+    """Obtain a dict mapping cached URLs to filenames."""
+    with _cache() as (dldir, url2hash):
+        return ReadOnlyDict((k, os.path.join(dldir, v))
+                            for (k, v) in url2hash.items())
+
+
+_cache_zip_index_name = "index.json"
+
+
+def export_download_cache(filename_or_obj, urls=None):
+    """Exports the cache contents as a ZIP file
+
+    Parameters
+    ----------
+    filename_or_obj : str or file-like
+        Where to put the created ZIP file. Must be something the zipfile
+        module can write to.
+    urls : iterable of str or None
+        The URLs to include in the exported cache. The default is all
+        URLs currently in the cache. If a URL is included in this list
+        but is not currently in the cache, it will be downloaded to the
+        cache first and then exported.
+
+    See Also
+    --------
+    import_download_cache : import the contents of such a ZIP file
+    import_to_cache : import a single file directly
+    """
+    if urls is None:
+        urls = get_cached_urls()
+    added = set()
+    with zipfile.ZipFile(filename_or_obj, 'w') as z:
+        index = {}
+        for u in urls:
+            fn = download_file(u, cache=True)
+            # Do not use os.path.join because ZIP files want
+            # "/" on all platforms
+            z_fn = "cache/" + os.path.basename(fn)
+            if z_fn not in added:
+                index[u] = z_fn
+                z.write(fn, z_fn)
+                added.add(z_fn)
+        z.writestr(_cache_zip_index_name, json.dumps(index))
+
+
+def import_download_cache(filename_or_obj, urls=None, update_cache=False):
+    """Imports the contents of a ZIP file into the cache
+
+    The ZIP file must be in the format produced by `export_download_cache`,
+    specifically it must have a file ``index.json`` that is a dictionary
+    mapping URLs to filenames inside the ZIP.
+
+    Parameters
+    ----------
+    filename_or_obj : str or file-like
+        Where to put the created ZIP file. Must be something the zipfile
```

Comment:
This sentence does not make sense. I thought the ZIP file is the input, not output?

Reply:
Oops.